pkg/util/webhookpatch.go

```
@@ -64,3 +75,107 @@ func PatchMutatingWebhookConfig(client admissionregistrationv1beta1client.Mutati
+
+				if oldConfig.ResourceVersion != newConfig.ResourceVersion {
+					for i, w := range newConfig.Webhooks {
+						if w.Name == webhookName && !bytes.Equal(newConfig.Webhooks[i].ClientConfig.CABundle, caCertPem) {
+							log.Infof("Detected a change in CABundle, patching MutatingWebhookConfiguration again")
+							shouldPatch <- struct{}{}
+							break
+						}
+					}
+				}
+			},
+		},
+	)
+	go controller.Run(stopCh)
+
+	go func() {
+		var delayedRetryC <-chan time.Time
+		if retry {
+			delayedRetryC = time.After(delayedRetryTime)
+		}
+
+		for {
+			select {
+			case <-delayedRetryC:
+				if retry := doPatch(client, injectionWebhookConfigName, webhookName, caCertPem); retry {
+					delayedRetryC = time.After(delayedRetryTime)
+				} else {
+					log.Infof("Retried patch succeeded")
+					delayedRetryC = nil
+				}
+			case <-shouldPatch:
+				if retry := doPatch(client, injectionWebhookConfigName, webhookName, caCertPem); retry {
+					if delayedRetryC == nil {
+						delayedRetryC = time.After(delayedRetryTime)
+					}
+				} else {
+					delayedRetryC = nil
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func doPatch(cs kubernetes.Interface, webhookConfigName, webhookName string, caCertPem []byte) (retry bool) {
+	client := cs.AdmissionregistrationV1beta1().MutatingWebhookConfigurations()
+	if err := PatchMutatingWebhookConfig(client, webhookConfigName, webhookName, caCertPem); err != nil {
+		log.Errorf("Patch webhook failed: %v", err)
+		return true
+	}
+	log.Infof("Patched webhook %s", webhookName)
+	return false
+}
+
+func CreateValidationWebhookController(client kubernetes.Interface, dynamicInterface dynamic.Interface,
+	webhookConfigName, ns, caBundlePath string) *controller.Controller {
+	o := controller.Options{
+		WatchedNamespace:  ns,
+		CAPath:            caBundlePath,
+		WebhookConfigName: webhookConfigName,
+		ServiceName:       "istiod",
+	}
+	whController, _ := controller.New(o, client, dynamicInterface)
```

why ignore the error here?

---

log the error now